contents/handbook/engineering/databases/clickhouse-event-table-migrations.md

```diff
@@ -0,0 +1,234 @@
+---
+title: Clickhouse event table migrations
+sidebar: Handbook
+showTitle: true
+---
+
+This document outlines how to do large-scale data migrations on cloud without using [Async Migrations](/handbook/engineering/databases/async-migrations).
+
+## Background
+
+Start of 2022 we [wanted to change events table schema to better support our querying patterns](https://github.com/PostHog/posthog/issues/5684).
+
+Doing this migration on cloud took several months and several false starts.
+
+## Migration strategy
+
+[Read guide to event ingestion before this](/handbook/engineering/databases/event-ingestion).
+
+Desired goals on the migration:
+1. It should be correct - no duplicated or missing data
+2. It should be timely - can be completed within a reasonable amount of time
+3. It should cause minimal ingestion delay
+4. It should keep materialized columns
+
+The rough migration strategy looks like this:
+
+<details><summary>1. Create a new staging table _without_ materialized columns on on 1 node on each of the shards.</summary>
+
+```sql
+CREATE TABLE posthog.sharded_events_ordered_by_event(
+    `uuid` UUID,
+    `event` String,
+    `properties` String,
+    `timestamp` DateTime64(6, 'UTC'),
+    `team_id` Int64,
+    `distinct_id` String,
+    `elements_hash` String,
+    `created_at` DateTime64(6, 'UTC'),
+    `_timestamp` DateTime,
+    `_offset` UInt64,
+    `elements_chain` String,
+    `$group_0` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_0'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
+    `$group_1` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_1'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
+    `$group_2` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_2'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
+    `$group_3` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_3'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
+    `$group_4` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$group_4'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
+    `$window_id` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$window_id'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), ''),
+    `$session_id` String MATERIALIZED replaceRegexpAll(JSONExtractRaw(properties, '$session_id'), concat('^[', regexpQuoteMeta('"'), ']*|[', regexpQuoteMeta('"'), ']*$'), '')
+)
+ENGINE = ReplicatedReplacingMergeTree(
+    '/clickhouse/prod/tables/{shard}/posthog.sharded_events_ordered_by_event3',
```

Comment:
```suggestion
    '/clickhouse/prod/tables/{shard}/posthog.sharded_events_ordered_by_event',
```

Reply:
I'm copying the raw SQL from my logs here - that was the zookeeper path used. Don't think we should replace this.